optaplanner-examples/src/test/java/org/optaplanner/examples/meetingscheduling/optional/score/MeetingSchedulingConstraintProviderTest.java

```diff
@@ -0,0 +1,616 @@
+
+        Meeting meeting2 = new Meeting();
+        meeting2.setDurationInGrains(4);
+
+        MeetingAssignment rightAssignment = new MeetingAssignment();
+        rightAssignment.setId(1L);
+        rightAssignment.setRoom(room);
+        rightAssignment.setStartingTimeGrain(timeGrain2);
+        rightAssignment.setMeeting(meeting2);
+
+        constraintVerifier.verifyThat(MeetingSchedulingConstraintProvider::roomConflict)
+                .given(leftAssignment, rightAssignment)
+                .penalizesBy(0);
+    }
+
+    @Test
+    public void roomConflictPenalized() {
+        Room room = new Room();
+
+        TimeGrain timeGrain1 = new TimeGrain();
+        timeGrain1.setGrainIndex(0);
+
+        Meeting meeting1 = new Meeting();
+        meeting1.setDurationInGrains(4);
+
+        MeetingAssignment leftAssignment = new MeetingAssignment();
+        leftAssignment.setId(0L);
+        leftAssignment.setRoom(room);
+        leftAssignment.setStartingTimeGrain(timeGrain1);
+        leftAssignment.setMeeting(meeting1);
+
+        TimeGrain timeGrain2 = new TimeGrain();
+        timeGrain2.setGrainIndex(2);
+
+        Meeting meeting2 = new Meeting();
+        meeting2.setDurationInGrains(4);
+
+        MeetingAssignment rightAssignment = new MeetingAssignment();
+        rightAssignment.setId(1L);
+        rightAssignment.setRoom(room);
+        rightAssignment.setStartingTimeGrain(timeGrain2);
+        rightAssignment.setMeeting(meeting2);
+
+        constraintVerifier.verifyThat(MeetingSchedulingConstraintProvider::roomConflict)
+                .given(leftAssignment, rightAssignment)
+                .penalizesBy(2);
+    }
+
+    @Test
+    public void avoidOvertimeUnpenalized() {
+        TimeGrain timeGrain = new TimeGrain();
+        timeGrain.setGrainIndex(3);
+
+        TimeGrain assignmentTimeGrain = new TimeGrain();
+        assignmentTimeGrain.setGrainIndex(0);
+
+        Meeting meeting = new Meeting();
+        meeting.setDurationInGrains(4);
+
+        MeetingAssignment meetingAssignment = new MeetingAssignment();
+        meetingAssignment.setStartingTimeGrain(assignmentTimeGrain);
+        meetingAssignment.setMeeting(meeting);
+
+        constraintVerifier.verifyThat(MeetingSchedulingConstraintProvider::avoidOvertime)
+                .given(meetingAssignment, timeGrain)
+                .penalizesBy(0);
+    }
+
+    @Test
+    public void avoidOvertimePenalized() {
+        TimeGrain assignmentTimeGrain = new TimeGrain();
+        assignmentTimeGrain.setGrainIndex(0);
+
+        Meeting meeting = new Meeting();
+        meeting.setDurationInGrains(4);
+
+        MeetingAssignment meetingAssignment = new MeetingAssignment();
+        meetingAssignment.setStartingTimeGrain(assignmentTimeGrain);
+        meetingAssignment.setMeeting(meeting);
+
+        constraintVerifier.verifyThat(MeetingSchedulingConstraintProvider::avoidOvertime)
+                .given(meetingAssignment)
+                .penalizesBy(3);
+    }
```

Comment:
The problem lies in `from(MeetingAssignment.class)`. There are no such instances, because you never set any value to `MeetingAssignment.room` planning variable. `from()` therefore treats those instances as uninitialized, and skips them. Two solutions:

- use `fromUnfiltered()` in the constraint provider or
- initialize your entities properly in the test.

Also, I suggest you make sure your `@PlanningId`s are never null, because the following PR will start enforcing that pretty soon:
https://github.com/triceo/optaplanner/pull/14

I have not reviewed any other constraints as I assumed the same problem would be plagueing them as well.

Reply:
I suggest you create constructors for the facts/entities so that you are forced to always provide all the information that must not be null.